src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -4429,60 +4403,121 @@ PartitionColumnIndex(Var *targetVar, List *targetList)
  */
 static StringInfo
 CreateMapQueryString(MapMergeJob *mapMergeJob, Task *filterTask,
-					 uint32 partitionColumnIndex)
+					 uint32 partitionColumnIndex, bool useBinaryFormat)
 {
 	uint64 jobId = filterTask->jobId;
 	uint32 taskId = filterTask->taskId;
+	char *resultNamePrefix = PartitionResultNamePrefix(jobId, taskId);
 
 	/* wrap repartition query string around filter query string */
 	StringInfo mapQueryString = makeStringInfo();
 	char *filterQueryString = TaskQueryString(filterTask);
-	char *filterQueryEscapedText = quote_literal_cstr(filterQueryString);
 	PartitionType partitionType = mapMergeJob->partitionType;
 
 	Var *partitionColumn = mapMergeJob->partitionColumn;
 	Oid partitionColumnType = partitionColumn->vartype;
-	char *partitionColumnTypeFullName = format_type_be_qualified(partitionColumnType);
-	int32 partitionColumnTypeMod = partitionColumn->vartypmod;
 
 	ShardInterval **intervalArray = mapMergeJob->sortedShardIntervalArray;
 	uint32 intervalCount = mapMergeJob->partitionCount;
 
 	if (partitionType == DUAL_HASH_PARTITION_TYPE)
 	{
 		partitionColumnType = INT4OID;
-		partitionColumnTypeMod = get_typmodin(INT4OID);
 		intervalArray = GenerateSyntheticShardIntervalArray(intervalCount);
 	}
 	else if (partitionType == SINGLE_HASH_PARTITION_TYPE)
 	{
 		partitionColumnType = INT4OID;
-		partitionColumnTypeMod = get_typmodin(INT4OID);
-	}
-
-	ArrayType *splitPointObject = SplitPointObject(intervalArray, intervalCount);
-	StringInfo splitPointString = ArrayObjectToString(splitPointObject,
-													  partitionColumnType,
-													  partitionColumnTypeMod);
-
-	char *partitionCommand = NULL;
-	if (partitionType == RANGE_PARTITION_TYPE)
-	{
-		partitionCommand = RANGE_PARTITION_COMMAND;
 	}
-	else
+	else if (partitionType == RANGE_PARTITION_TYPE)
 	{
-		partitionCommand = HASH_PARTITION_COMMAND;
+		/* add a partition for NULL values at index 0 */
+		intervalArray = RangeIntervalArrayWithNullBucket(intervalArray, intervalCount);
+		intervalCount++;
```

Comment:
We have logic in MergeTaskList that expects an additional bucket to be generated for cases where the join column value is NULL. This was previously done implicitly in worker_range_partition_table, but it seems cleaner to define an actual bucket in the interval array.

We may need to revise this when we allow NULL values in range-partitioned tables, since there is already a bucket in that case (see comments in MergeTaskList and RangeIntervalArrayWithNullBucket).

Alternatively, we could consider removing range-re-partitioning, but it might come in handy one day.